File: src/python/graph/4_get_branch.py
# file: 4_get_branch.py
# Extract graph branch starting from node, stopping at chips/common nets

import networkx as nx # read_gml, write_gml
import sys            # argv, exit

def is_chip(G, node):
    if G.nodes[node].get('type') != 'component': return False
    return sum(1 for c in node if c.isalpha() and c.isascii()) > 1

def traverse(G, node, common_nets, max_nodes, visited=None):
    if visited is None: visited = set()
    if max_nodes > 0 and len(visited) >= max_nodes: return []
    visited.add(node)
    edges = []
    for nbr in G.neighbors(node):
        if nbr not in visited:
            if max_nodes > 0 and len(visited) >= max_nodes: break
            edges.append((node, nbr))
            if nbr not in common_nets and not is_chip(G, nbr):
                edges.extend(traverse(G, nbr, common_nets, max_nodes, visited))
    return edges

def main():
    if len(sys.argv) < 5: sys.exit(1)
    start, input_gml, output_gml, common_file = sys.argv[1:5]
    max_nodes = int(sys.argv[5]) if len(sys.argv) > 5 else 25
    with open(common_file, 'r') as f:
        common_nets = set(line.strip() for line in f if line.strip())
    try: G = nx.read_gml(input_gml)
    except FileNotFoundError: sys.exit(0)
    if start not in G:
        with open(output_gml, "w") as f: f.write("NOT_CONNECTED")
        sys.exit(1)
    edges = traverse(G, start, common_nets, max_nodes)
    H = nx.Graph()
    H.add_edges_from(edges)
    for node in H.nodes():
        if node in G.nodes: H.nodes[node].update(G.nodes[node])
    nx.write_gml(H, output_gml)

if __name__ == "__main__":
    main()

File: src/typescript/graph/4_get_branch.ts
import fs from 'fs';
import Graph from 'graphology';

function isChip(G: Graph, node: string): boolean {
    if (G.getNodeAttributes(node).type !== 'component') return false;
    return (node.match(/[a-zA-Z]/g)?.length || 0) > 1;
}

function traverse(G: Graph, node: string, commonNets: Set<string>, maxNodes: number, visited: Set<string>): { u: string, v: string }[] {
    if (maxNodes > 0 && visited.size >= maxNodes) return [];
    visited.add(node);
    const edges: { u: string, v: string }[] = [];
    
    G.neighbors(node).forEach(nbr => {
        if (!visited.has(nbr)) {
            if (maxNodes > 0 && visited.size >= maxNodes) return;
            edges.push({ u: node, v: nbr });
            if (!commonNets.has(nbr) && !isChip(G, nbr)) {
                edges.push(...traverse(G, nbr, commonNets, max_nodes, visited));
            }
        }
    });
    return edges;
}

// Keeping version compatible with what 7_text_appender expects but also mirroring 4_get_branch.py logic
export function getBranch(G: Graph, startNode: string, commonNets: Set<string>, maxNodes: number = 25): Graph {
    const H = new Graph();
    if (!G.hasNode(startNode)) return H;

    const visited = new Set<string>();
    const edges = traverse(G, startNode, commonNets, max_nodes, visited);
    
    H.addNode(startNode, G.getNodeAttributes(startNode));
    edges.forEach(e => {
        if (!H.hasNode(e.u)) H.addNode(e.u, G.getNodeAttributes(e.u));
        if (!H.hasNode(e.v)) H.addNode(e.v, G.getNodeAttributes(e.v));
        H.addEdge(e.u, e.v);
    });
    
    return H;
}
